tests/periph_uart_mode/main.c

```diff
@@ -0,0 +1,153 @@
+ * shell. This test will cycle through all permutations of UART modes
+ * and print the mode to STDOUT. The test is evaluated by hooking up
+ * a probe or logic analyzer and confirming that each string can be
+ * read correctly. The strings will be printed as such:
+ * <data-bits><parity-bits><stop-bits>
+ * Example: 6O1 would be 6 data bits, odd parity, 1 stop bit
+ * If a mode is not supported, then a message indicating that will
+ * be printed using configuration from the last mode);
+ */
+
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+
+#include "periph/uart.h"
+#include "xtimer.h"
+
+/* UART_DEV is always 0 since this test is for device with 1 UART */
+#define _UART_DEV (UART_DEV(0))
+
+
+static void _print_status(int8_t status, uart_data_bits_t data_bits,
+                          uart_parity_t parity, uart_stop_bits_t stop_bits) {
+    char mode[4];
+
+    /* If status < 0, print a header indicating an error */
+    if (status < 0) {
+        printf("Unsupported Mode: ");
+    }
+
+    switch (data_bits) {
+        case UART_DATA_BITS_5:
+            mode[0] = '5';
+            break;
+        case UART_DATA_BITS_6:
+            mode[0] = '6';
+            break;
+        case UART_DATA_BITS_7:
+            mode[0] = '7';
+            break;
+        case UART_DATA_BITS_8:
+            mode[0] = '8';
+            break;
+        default:
+            break;
+    }
+
+    switch (parity) {
+        case UART_PARITY_NONE:
+            mode[1] = 'N';
+            break;
+        case UART_PARITY_EVEN:
+            mode[1] = 'E';
+            break;
+        case UART_PARITY_ODD:
+            mode[1] = 'O';
+            break;
+        case UART_PARITY_MARK:
+            mode[1] = 'M';
+            break;
+        case UART_PARITY_SPACE:
+            mode[1] = 'S';
+            break;
+        default:
+            break;
+    }
+
+    switch (stop_bits) {
+        case UART_STOP_BITS_1:
+            mode[2] = '1';
+            break;
+        case UART_STOP_BITS_2:
+            mode[2] = '2';
+            break;
+        default:
+            break;
+    }
+
+    mode[3] = '\0';
+
+    printf("%s\n", mode);
+}
+
+
+int main(void)
+{
+
+    /* Use arrays so that each option can be easily iterated over */
+    uart_data_bits_t data_bits[4] = {
+        UART_DATA_BITS_5,
+        UART_DATA_BITS_6,
+        UART_DATA_BITS_7,
+        UART_DATA_BITS_8
+    };
+
+    uart_parity_t parity[5] = {
+        UART_PARITY_NONE,
+        UART_PARITY_EVEN,
+        UART_PARITY_ODD,
+        UART_PARITY_MARK,
+        UART_PARITY_SPACE
+    };
+
+    uart_stop_bits_t stop_bits[2] = {
+        UART_STOP_BITS_1,
+        UART_STOP_BITS_2
+    };
+
+    /* Test each permutation */
+    for (int didx = 0; didx < 4; ++didx) {
+        for (int pidx = 0; pidx < 5; ++pidx) {
+            for (int sidx = 0; sidx < 2; ++sidx) {
+                int8_t status = uart_mode(_UART_DEV, data_bits[didx],
```

Comment:
Though this implementation is not affected by this, in the api it says to always init before changing mode so you will have to call `uart_init(_UART_DEV, <BAUDRATE>, NULL, NULL);`

Reply:
Maybe you can also hardcode (or overwritable define) the baud if you want..  9600 would cover all worst case assuming the scope/logic analyzer has enough memory.

Reply:
Since I've been using stdout for these, would stdio_init be OK? Or I could switch to calling uart_init directly and then just using uart_write() instead. Either one works for me.

Reply:
Since I've been using stdout, the baud can be set at compile-time with `CFLAGS += -DSTDIO_UART_BAUDRATE=<BAUD>`

Reply:
I'm very close to being indifferent.  I said uart_init just because that what was discussed in the initial implementation PR #10743 
However I don't see it in the API documentation... oops.

Reply:
yes, I thought that the requirement was just that the mode switch was done after a uart_init() had been done, not that a uart_init() was needed for every mode change. Happy to make this change if need be, but I do not think it is necessary.

Reply:
In the STM implementation it assumes that the mode registers are empty.  I think it ors without first anding (save those precious bytes).  So it would be good to have it in the test for now.  Thanks!

Reply:
Sounds good, kept with stdio_init() but calling it before each mode switch now.

Reply:
ah, forgot to put the note in the README about baud, just added and re-pushed